src/coreclr/jit/lclvars.cpp

```diff
@@ -4034,17 +4034,22 @@ size_t LclVarDsc::lvArgStackSize() const
 //
 var_types LclVarDsc::GetRegisterType(const GenTreeLclVarCommon* tree) const
 {
-    var_types targetType = tree->gtType;
-    var_types lclVarType = TypeGet();
+    var_types targetType = tree->TypeGet();
 
     if (targetType == TYP_STRUCT)
     {
-        if (lclVarType == TYP_STRUCT)
+        ClassLayout* layout;
+        if (tree->OperIs(GT_LCL_FLD, GT_STORE_LCL_FLD))
         {
-            assert(!tree->OperIsLocalField() && "do not expect struct local fields.");
-            lclVarType = GetLayout()->GetRegisterType();
+            layout = tree->AsLclFld()->GetLayout();
```

Comment:
The documentation on this function seems a little misleading now. Generally it feels like the function should maybe be renamed something like `GetRegisterTypeForAccess`?

Reply:
Yes. I guess most of all it seems like a weird inversion to me that this is a function on `LclVarDsc` when the local  is completely determined by the local node.

Let's leave the potential cleanup to a future change since this handled primitively typed LCL_FLD nodes before as well.

Reply:
> I guess most of all it seems like a weird inversion to me that this is a function on LclVarDsc when the local is completely determined by the local node.

Ah yes, this is because we don't have access to the `LclVarDsc*` from inside `GenTree*`. The more idiomatic pattern for this problem in the code is to have `GenTree::NeedsLclVarDsc(Compiler* compiler)`, here it was put on `LclVarDsc*` instead, I suppose because it was already available at most callsites.

> Let's leave the potential cleanup to a future change since this handled primitively typed LCL_FLD nodes before as well.

Added to my TODO list.